### ISSUE TITLE
BUG: what if LDBL_EPSILON is zero (like on ppc64le)?

### ISSUE THREAD

**Issue by mattip** [numpy/numpy#15763]
On some systems, `LDBL_EPSILON`, which is defined as a number that fits `(1.0L + LDBL_EPSILON) != 1.0L && (1.0L + LDBL_EPSILON/2) == 1.0L`, is 0. This causes problems in [complex math routines](https://github.com/numpy/numpy/blob/v1.18.1/numpy/core/src/npymath/npy_math_complex.c.src#L60), especially when calculating the [recipricol](https://github.com/numpy/numpy/blob/v1.18.1/numpy/core/src/npymath/npy_math_complex.c.src#L1408) as `const npy_longdouble RECIP_EPSILON = 1.0L / LDBL_EPSILON`;

I discovered this when running `-mfull` tests on ppc64le. I got an overflow warning when calling `np.arcsinh`, and debugging led to the conclusion that `LDBL_EPSILON` is `0`.

One possible mitigation is to add `RECIP_EPSILON` to the pre-calculated values on the line above it, but that presumes that the internal representation of longdouble, which [wikipedia says](https://en.wikipedia.org/wiki/Long_double) may be problematic. On the other hand, all the values in these calculations seem to presume a X86 80-bit long double.

Thoughts?

**Comment by seberg** [comment]
Hmm, but LDBL_EPSILON being 0 is just incorrect, right? What does `np.finfo(np.longdouble).eps` say on those machines? I am not sure the way that is detected is available at compile time easily, but maybe we can make it work.
Agreed about the other constants, they should be different on different platforms, not sure what a good solution is, unless you can somehow calculate those values as a static based on the corrected eps value?

**Comment by charris** [comment]
IIRC, we hardwired in a value for that number. What you see is probably an oversight.

**Comment by charris** [comment]
In `getlimits.py`
```
    float_dd_ma = MachArLike(ld,
                              machep=-105,
                              negep=-106,
                              minexp=-1022,
                              maxexp=1024,
                              it=105,
                              iexp=11,
                              ibeta=2,
                              irnd=5,
                              ngrd=0,
                              eps=exp2(ld(-105)),
                              epsneg= exp2(ld(-106)),
                              huge=huge_dd,
                              tiny=exp2(ld(-1022)))
```

**Comment by seberg** [comment]
@charris but those are the limits we use on the Python side during runtime, on the c-side `LDBL_EPSILON` is defined in `float.h` by gcc (at least on my system)... Should we try to define our own `NPY_LDBL_EPSILON` to fall back to the `getlimits.py` definition? I am not quite remember the story behind the new `getlimits.py` definitions compared to the C-side `LDBL_EPSILON` for example. Or is t here some runtime floating point mode difference?

**Comment by mattip** [comment]
> but LDBL_EPSILON being 0 is just incorrect, right? 

I found very few references to it, but the ones I found did say it can be 0

I am defining it hard-coded as the 80-bit long version, let's see if there are problems with tests failing.

**Comment by mattip** [comment]
This is on the gcc build farm with a redhat 7.7.1908 OS using gcc 4.8.5. The travis build machine does not have these problems, I wonder what gcc it uses. Anyway, perhaps this is irrelevant for the wheels since we will be building them with gcc 8.3.1.

When I do 
```
diff --git a/numpy/core/src/npymath/npy_math_complex.c.src b/numpy/core/src/npymath/npy_math_complex.c.src
index 8c432e4..6dcdf8e 100644
--- a/numpy/core/src/npymath/npy_math_complex.c.src
+++ b/numpy/core/src/npymath/npy_math_complex.c.src
@@ -49,6 +49,8 @@
 static const volatile npy_float tiny = 3.9443045e-31f;
 #endif
 
+#define NPY_LDBL_EPSILON 1.08420217248550443400745280086994171e-19L
+
 /**begin repeat
  * #type = npy_float, npy_double, npy_longdouble#
  * #ctype = npy_cfloat,npy_cdouble,npy_clongdouble#
@@ -57,7 +59,7 @@ static const volatile npy_float tiny = 3.9443045e-31f;
  * #TMAX = FLT_MAX, DBL_MAX, LDBL_MAX#
  * #TMIN = FLT_MIN, DBL_MIN, LDBL_MIN#
  * #TMANT_DIG = FLT_MANT_DIG, DBL_MANT_DIG, LDBL_MANT_DIG#
- * #TEPS = FLT_EPSILON, DBL_EPSILON, LDBL_EPSILON#
+ * #TEPS = FLT_EPSILON, DBL_EPSILON, NPY_LDBL_EPSILON#
  * #precision = 1, 2, 3#
  */
```

Then there is one failing test in `TestComplexFunctions.test_loss_of_precision[complex256]`

```
x = array([1.00000000e-20, 1.21736864e-20, 1.48198641e-20, 1.80412378e-20,
       2.19628372e-20, 2.67368693e-20, 3.254862...5156e-04, 4.54267386e-04,
       5.53010871e-04, 6.73218092e-04, 8.19554595e-04, 9.97700064e-04],
      dtype=float128)
rtol = 1.23259516440783094595582588325435e-30

    def check(x, rtol):
        x = x.astype(real_dtype)
    
        z = x.astype(dtype)
        d = np.absolute(np.arcsinh(x)/np.arcsinh(z).real - 1)
        assert_(np.all(d < rtol), (np.argmax(d), x[np.argmax(d)], d.max(),
>                                 'arcsinh'))
E       AssertionError: (131, 1.54987431981169249551435343964035e-09, 4.0035173453529620614007210953362e-19, 'arcsinh')
```

**Comment by mattip** [comment]
> What does np.finfo(np.longdouble).eps say on those machines? 

```
>>> np.finfo(np.longdouble).eps
2.4651903288156618919116517665087e-32
>>> np.finfo(np.longdouble)
finfo(resolution=1e-31, min=-1.79769313486231580793728971405301e+308, \
 max=1.79769313486231580793728971405301e+308, dtype=float128)
```

**Comment by seberg** [comment]
@mattip this is probably very silly, but what does this print on that machine:
```
#define STR(x) #x
#define GET_DEFINE(x) STR(x)
    printf("*********************************************\n");
    printf("Longdouble macro definition: " GET_DEFINE(LDBL_EPSILON) "\n");
    printf("*********************************************\n");
```
I saw one stackoverflow saying that the `L` was missing on some compilers, in which case maybe changing some of the other constants (e.g. the integers) in that file might help?

**Comment by mattip** [comment]
```
*********************************************
Longdouble macro definition: 4.94065645841246544176568792868221e-324L
*********************************************
```

which seems to be way too small. We don't see this with the travis machines, let's see if it shows up on the numpy-wheel builds (which are crashing for other reasons) with gcc 8.3.1.

**Comment by seberg** [comment]
That is a weird number, looks more like the smallest subnormal than a reasonable value. To be honest, that seems like a compiler bug, no?

**Comment by seberg** [comment]
Hmmm, wait, your `finfo` is a bit strange though as well? The min/max are identical to double, but the eps is smaller than the extended precision eps of `1.084202172485504434e-19`? What is long double on those machines to begin with? Or is it just double and the generic code to figure out the `eps` used by `getlimits.py` is wrong and forgets about rounding?

**Comment by charris** [comment]
> I am not quite remember the story behind the new getlimits.py definitions 

We needed to fix it in order to support PPC. It used to be computed but didn't work because the computation assumed the usual mantissa/exponent representation. The number used was recommended somewhere (IBM?). @matthew-brett was responsible for the actual work in #8504.

> What is long double on those machines to begin with?

IBM DoubleDouble, or should be.

**Comment by seberg** [comment]
Ah OK, double-double is defined as the sum of two double precision numbers, so the range and eps are a bit strange I guess, and do not follow normal floating point logic. So that makes sense then, the `eps` is just defned as `np.nextafter(0., 1.)` for the single double.

But, it also makes sense that double-double arithmic should be using a different eps value, I am not sure what exactly, but I suppose something more like `np.finfo(np.double).eps**2` may be reasonable?

**Comment by charris** [comment]
ISTR that POWER9 has gone to IEEE quad precision, so we might want to be careful. There are macros defined in `numpy/core/setup_common.py` for the different floating types.

**Comment by Flamefire** [comment]
I found the related PR https://github.com/numpy/numpy/pull/16214 where @seberg even wrote:

> Probably broken for all double-double 

I tested the `bad_arcsin` on our PPC system and there it succeeded, although the test failed.
TLDR: Replacing the constant  by any of `[2.16808037e-10, 2.63935305e-10, 3.21306564e-10, 3.91148536e-10, 4.76171961e-10, 5.79676814e-10, 7.05680376e-10, 8.59073160e-10, 1.04580873e-09, 1.27313475e-09, 1.54987432e-09]` fails the same test so that can be used on PPC (POWER9) too.

Detailed analysis:

```
import numpy as np
dtype=np.longcomplex
real_dtype = dtype(0.).real.dtype
eps = np.finfo(dtype).eps
x = np.logspace(-20, -3.001, 200).astype(real_dtype)
z = x.astype(dtype)
d = np.absolute(np.arcsinh(x)/np.arcsinh(z).real - 1)
d[d >= 50*eps]
x[d >= 50*eps]
```

The last 2 lines yield:
```
array([7.83428747e-21, 1.16103076e-20, 1.72063180e-20, 2.54995295e-20,
       3.77899561e-20, 5.60042014e-20, 8.29974654e-20, 1.23001116e-19,
       1.82285982e-19, 2.70145348e-19, 4.00351735e-19], dtype=float128)
array([2.16808037e-10, 2.63935305e-10, 3.21306564e-10, 3.91148536e-10,
       4.76171961e-10, 5.79676814e-10, 7.05680376e-10, 8.59073160e-10,
       1.04580873e-09, 1.27313475e-09, 1.54987432e-09], dtype=float128)
```


GLIBC 2.17 used here, no option to use any other as this is an IBM maintained (at the core) HPC system.

Similar the test code, adapted to another value:
```
import numpy as np
x=1.54987432e-09
v1 = np.arcsinh(np.float128(x))
v2 = np.arcsinh(np.complex256(x)).real
(v1 / v2) - 1.0
```
--> `4.0035173463258060912440733481724e-19` and `<stdin>:1: RuntimeWarning: overflow encountered in arcsinh` on the `v2` line

The overflow warning comes for all values of the above list I tried, so that might be the reason for the inaccuracy.

Anyway: I propose to choose another number which works (i.e. fails) for both PPC and AArch64 and skip the test on PPC the same way as on AArch64.

**Comment by Flamefire** [comment]
This is still an issue (numpy 1.22.3)
